[win/asan] Support instructions in GetInstructionSize used by Wine.

bernhardu opened the pull request:
This patch adds several instructions seen when trying to run a executable built with ASan with llvm-mingw,
with the tip of the main branch from llvm-project (x86 and x86_64). Also another missing instruction visible in the issue 96270.

Also includes instructions collected by
Roman Pišl and Eric Pouech in the Wine bug reports below.

~~Also changes "44 0f b6 1a" to return 4 instead of 5 (MR 111638).~~ Got fixed in 8417f6af54c.

Fixes: https://github.com/llvm/llvm-project/issues/96270

Co-authored-by: Roman Pišl
                https://bugs.winehq.org/show_bug.cgi?id=50993
                https://bugs.winehq.org/attachment.cgi?id=70233

Co-authored-by: Eric Pouech
                https://bugs.winehq.org/show_bug.cgi?id=52386
                https://bugs.winehq.org/attachment.cgi?id=71626

bernhardu commented:
v2:
- Added another instruction "81 ec XX XX XX XX" which showed up using clang to compile the 32-bit PE side of Wine.

bernhardu commented:
v3:
- Another few instructions, which showed up after getting the wine build working, with the 64-bit clang PE side.

bernhardu commented:
v4:
- rebase over the recent fix 8417f6a
- add a few more instructions found when trying more applications.

@zmodem Thanks for the offer, and sorry for the delay.
I wonder if this has chances to go in, as it targets not a native windows platform?
And for "41 b8 XX XX XX XX : mov r8d, XX XX XX XX" I am not sure if it should
also consider rel_offset, but I don't know how to test this.

bernhardu commented:
v5:
- Rebase to current tip.
- Improve commit comment (hopefully).

bernhardu commented:
Unfortunately just found and will change in next version:
- `83 E4 F8 : and esp, 0xFFFFFFF8` and `83 e4 e0 : and esp, 0xFFFFFFE0`
  should be already handled by `83 E4 XX : and esp, XX`

bernhardu commented:
v6:
- split off the changes modifying existing lines in separate commits.

bernhardu commented:
v7:
- Found with the test a duplicate instruction, removed the line which got never reached.
- Split the test addition into different commits.

Sorry for the ongoing changes, please drop me a note if some commits should go into a separate PR.

rnk reviewed (commented):
@zmodem, are you able to review this at some point?

mstorsjo commented on compiler-rt/lib/interception/tests/interception_win_test.cpp:
```diff
@@ -230,6 +230,7 @@ const u8 kUnpatchableCode6[] = {
     0x90, 0x90, 0x90, 0x90,
 };
 
+#if SANITIZER_WINDOWS64
```
Why these new `#if`s here (same for the others below)? So far, these tests pass fine on both i686 and x86_64, see e.g. https://github.com/mstorsjo/llvm-mingw/actions/runs/11883858793/job/33120364287.

mstorsjo replied:
Ah, I see that this gets rid of warnings like this: `warning: unused variable 'kUnpatchableCode7' [-Wunused-const-variable]`

That's probably fine, but it would distract less from the rest of this MR if it would be filed as a separate cleanup.

mstorsjo commented on compiler-rt/lib/interception/interception_win.cpp:
```diff
@@ -658,7 +681,6 @@ static size_t GetInstructionSize(uptr address, size_t* rel_offset = nullptr) {
   }
 
   switch (0x00FFFFFF & *(u32*)address) {
-    case 0x07c1f6:    // f6 c1 07 : test cl, 0x7
```
Why are some old existing cases removed?

bernhardu replied:
That `f6 c1 07 : test cl, 0x7` got moved into the generic form `F6 C1 XX : test cl, XX`.
I tried to document this by moving it into separate commits before the "big" one.

mstorsjo replied:
Ah, I see - so this should ideally be reviewed as a series of individual patches, rather than looking at the total result of this PR. The policy within llvm is (IMO unfortunately) that PRs are merged with "squash and merge" only, so the incoming distinction between individual commits is lost on merge, and most reviewers probably don't look at it with a per-commit view (unless told to).

mstorsjo reviewed (commented):
If this interacts with https://github.com/llvm/llvm-project/commit/8417f6af54c8f6dcf5893ab1352b50bf33c5a1ba, can you rebase this change past that, to make things clearer with how it fits together with that?

bernhardu commented:
> If this interacts with [8417f6a](https://github.com/llvm/llvm-project/commit/8417f6af54c8f6dcf5893ab1352b50bf33c5a1ba), can you rebase this change past that, to make things clearer with how it fits together with that?

Since the comment "v4" it is already rebased behind 8417f6af, so this should make no problems.
But I see now there was another commit to this file in cdfd4cf. It seems to make no conflict, but should I still rebase over this one?

Thanks for having a look.
So should I mark this PR as draft again, and feed in separate PRs for easier review?

mstorsjo commented:
> > If this interacts with [8417f6a](https://github.com/llvm/llvm-project/commit/8417f6af54c8f6dcf5893ab1352b50bf33c5a1ba), can you rebase this change past that, to make things clearer with how it fits together with that?
> 
> Since the comment "v4" it is already rebased behind [8417f6a](https://github.com/llvm/llvm-project/commit/8417f6af54c8f6dcf5893ab1352b50bf33c5a1ba), so this should make no problems. But I see now there was another commit to this file in [cdfd4cf](https://github.com/llvm/llvm-project/commit/cdfd4cff555db532a9001662fc8d6a7b05f1c2c3). It seems to make no conflict, but should I still rebase over this one?
> 
> Thanks for having a look. So should I mark this PR as draft again, and feed in separate PRs for easier review?

Ah, ok, I see.

As I'm not qualified enough to really review this, I would hold off of updating it until some of those who might be (@rnk, @zmodem, @vitalybuka or @barcharcraz) says which way they prefer to handle it.

As for adding the extra `#if`s to silence warnings due to the unused code blocks - just file that in a separate PR which I can accept quite quickly.

bernhardu commented:
> ... The policy within llvm is ... that PRs are merged with "squash and merge" ...

Oh, I was not aware of this, so this seems also to suggest to use the separate PR approach, Thanks.

alvinhochun commented on compiler-rt/lib/interception/interception_win.cpp:
```diff
@@ -643,6 +664,8 @@ static size_t GetInstructionSize(uptr address, size_t* rel_offset = nullptr) {
     case 0x7980:  // 80 79 YY XX  cmp BYTE ptr [rcx+YY], XX
       return 4;
 
+    case 0xB841:  // 41 B8 XX XX XX XX : mov r8d, XX XX XX XX
```
This looks wrong. The second operand is an immediate, not a relative offset, so it shouldn't be leading to a branch that sets `rel_offset`.

alvinhochun commented on compiler-rt/lib/interception/interception_win.cpp:
```diff
@@ -634,7 +656,6 @@ static size_t GetInstructionSize(uptr address, size_t* rel_offset = nullptr) {
     case 0x018a:  // mov al, byte ptr [rcx]
       return 2;
 
-    case 0x058A:  // 8A 05 XX XX XX XX : mov al, byte ptr [XX XX XX XX]
```
Just want to note that this line is being moved to below (line 668), which fixes a wrong change from https://github.com/llvm/llvm-project/commit/3bd8f4e0a05f2eecc2029ff05d21f19e74b6eb67#diff-ca7181301de35b7cb405b0809770e12d68a5466b396c7f0e5cf61ab2145c828dR637-R649 .

It may be worth making a separate review for this single change (or even just committing the fix directly) to have it fixed quickly.

bernhardu replied:
Splitted of this PR: GetInstructionSize: Fix `8A 05 ...` to return 6 again.
https://github.com/llvm/llvm-project/pull/116889
Thank you, hope it is ok when I resolve this paragraph.

zmodem reviewed (commented):
Sorry for the delay, I'll try to prioritize this.

I did a first pass, but didn't spot check any of the actual encodings yet. If you do have some parts that could be separate PRs, that would be helpful.

bernhardu commented:
Thank you very much for the great feedback.
I started with splitting of this PR:
https://github.com/llvm/llvm-project/pull/116887

bernhardu commented:
Splitted of this PR: GetInstructionSize: Fix `8A 05 ...` to return 6 again.
https://github.com/llvm/llvm-project/pull/116889

bernhardu commented:
Splitted of this PR: Remove duplicate instruction `FF 25 ....`
https://github.com/llvm/llvm-project/pull/116894

bernhardu commented on compiler-rt/lib/interception/interception_win.cpp:
```diff
@@ -798,11 +872,9 @@ static size_t GetInstructionSize(uptr address, size_t* rel_offset = nullptr) {
     case 0x75FF:  // FF 75 XX : push dword ptr [ebp + XX]
       return 3;
     case 0xC1F7:  // F7 C1 XX YY ZZ WW : test ecx, WWZZYYXX
-    case 0x25FF:  // FF 25 XX YY ZZ WW : jmp dword ptr ds:[WWZZYYXX]
```
I found this when I created the test.
There is a similar line some lines above which returns 0 as failure for a jmp instruction.

Splitted of this PR: Remove duplicate instruction `FF 25 ....`
https://github.com/llvm/llvm-project/pull/116894

bernhardu commented on compiler-rt/lib/interception/tests/interception_win_test.cpp:
```diff
@@ -793,6 +799,294 @@ TEST(Interception, EmptyExportTable) {
+    u8 instr[16];
+    SIZE_T rel_offset;
+  } data[] = {
+    /* sort list */
```
Unfortunately not, I was basically just copying GetInstructionSize
and editing with notepad++`s multi-line-edit and macro feature.

bernhardu commented:
Splitted of this PR: Add a test skeleton for function GetInstructionSize.
https://github.com/llvm/llvm-project/pull/116948

bernhardu commented on compiler-rt/lib/interception/interception_win.h:
```diff
@@ -63,6 +63,9 @@ bool OverrideFunctionWithTrampoline(
 void TestOnlyReleaseTrampolineRegions();
 
+// Exposed for unittests
+SIZE_T test_GetInstructionSize(uptr address, SIZE_T* rel_offset);
```
Done in the splitted PR.

bernhardu commented on compiler-rt/lib/interception/tests/interception_win_test.cpp:
```diff
@@ -793,6 +799,294 @@ TEST(Interception, EmptyExportTable) {
+TEST(Interception, GetInstructionSize) {
+
+  struct {
+    SIZE_T size;
```
Done in the splitted PR.

bernhardu commented on compiler-rt/lib/interception/tests/interception_win_test.cpp:
```diff
@@ -793,6 +799,294 @@ TEST(Interception, EmptyExportTable) {
+    SIZE_T rel_offset;
+  } data[] = {
+    /* sort list */
+    {  0, { 0x70, 0x71 }, 0 },  // 70 71 : jo XX (short conditional jump)
```
Done in the splitted PR.

bernhardu commented on compiler-rt/lib/interception/tests/interception_win_test.cpp:
```diff
@@ -793,6 +799,294 @@ TEST(Interception, EmptyExportTable) {
+    size = __interception::test_GetInstructionSize((uptr)data[i].instr, &rel_offset);
+    EXPECT_EQ(data[i].size, size)
+            << "  with i=" << i << " ( "
+            << std::setfill('0') << std::setw(2) << std::right << std::hex << (int)data[i].instr[0] << " "
+            << std::setfill('0') << std::setw(2) << std::right << std::hex << (int)data[i].instr[1] << " "
+            << std::setfill('0') << std::setw(2) << std::right << std::hex << (int)data[i].instr[2] << " "
```
Done in the splitted PR.

bernhardu committed:
TEMP: Import minimal copy of bddisasm for evaluation.

bernhardu committed:
TEMP: .gitignore: Exclude build directories.

bernhardu committed:
TEMP: Allow to run test at i686.

At least in Wine.

bernhardu committed:
TEMP: Add debugging output.

bernhardu committed:
TEMP: Use bddisasm to verify test results.

bernhardu committed:
[win/asan] GetInstructionSize: Support some more 3 byte instructions.

This patch adds several instructions seen when trying to run a
executable built with ASan with llvm-mingw.
(x86 and x86_64, using the git tip in llvm-project).

Also includes instructions collected by
Roman Pišl and Eric Pouech in the Wine bug reports below.

Related: https://github.com/llvm/llvm-project/issues/96270

Co-authored-by: Roman Pišl <rpisl@seznam.cz>
                https://bugs.winehq.org/show_bug.cgi?id=50993
                https://bugs.winehq.org/attachment.cgi?id=70233
Co-authored-by: Eric Pouech <eric.pouech@gmail.com>
                https://bugs.winehq.org/show_bug.cgi?id=52386
                https://bugs.winehq.org/attachment.cgi?id=71626

bernhardu committed:
[win/asan] GetInstructionSize: Support some more 7 or 8 byte instructions.

This patch adds several instructions seen when trying to run a
executable built with ASan with llvm-mingw.
(x86 and x86_64, using the git tip in llvm-project).

Also includes instructions collected by
Roman Pišl and Eric Pouech in the Wine bug reports below.

Related: https://github.com/llvm/llvm-project/issues/96270

Co-authored-by: Roman Pišl <rpisl@seznam.cz>
                https://bugs.winehq.org/show_bug.cgi?id=50993
                https://bugs.winehq.org/attachment.cgi?id=70233
Co-authored-by: Eric Pouech <eric.pouech@gmail.com>
                https://bugs.winehq.org/show_bug.cgi?id=52386
                https://bugs.winehq.org/attachment.cgi?id=71626

bernhardu committed:
[win/asan] GetInstructionSize: Support some more 10 or more byte instructions.

This patch adds several instructions seen when trying to run a
executable built with ASan with llvm-mingw.
(x86 and x86_64, using the git tip in llvm-project).

Also includes instructions collected by
Roman Pišl and Eric Pouech in the Wine bug reports below.

Fixes: https://github.com/llvm/llvm-project/issues/96270

Co-authored-by: Roman Pišl <rpisl@seznam.cz>
                https://bugs.winehq.org/show_bug.cgi?id=50993
                https://bugs.winehq.org/attachment.cgi?id=70233
Co-authored-by: Eric Pouech <eric.pouech@gmail.com>
                https://bugs.winehq.org/show_bug.cgi?id=52386
                https://bugs.winehq.org/attachment.cgi?id=71626

bernhardu committed:
[win/asan] GetInstructionSize: Support some more 4 byte instructions.

This patch adds several instructions seen when trying to run a
executable built with ASan with llvm-mingw.
(x86 and x86_64, using the git tip in llvm-project).

Also includes instructions collected by
Roman Pišl and Eric Pouech in the Wine bug reports below.

Related: https://github.com/llvm/llvm-project/issues/96270

Co-authored-by: Roman Pišl <rpisl@seznam.cz>
                https://bugs.winehq.org/show_bug.cgi?id=50993
                https://bugs.winehq.org/attachment.cgi?id=70233
Co-authored-by: Eric Pouech <eric.pouech@gmail.com>
                https://bugs.winehq.org/show_bug.cgi?id=52386
                https://bugs.winehq.org/attachment.cgi?id=71626

bernhardu committed:
TEMP: Enable interception of Heap(Alloc|Free|ReAlloc|Size) in Wine.

Wine-Bug: https://bugs.winehq.org/show_bug.cgi?id=57498

bernhardu committed:
[win/asan] GetInstructionSize: Add test for `8D A4 24 ...`.

bernhardu committed:
[win/asan] GetInstructionSize: Support some more 2 byte instructions.

This patch adds several instructions seen when trying to run a
executable built with ASan with llvm-mingw.
(x86 and x86_64, using the git tip in llvm-project).

Also includes instructions collected by
Roman Pišl and Eric Pouech in the Wine bug reports below.

Related: https://github.com/llvm/llvm-project/issues/96270

Co-authored-by: Roman Pišl <rpisl@seznam.cz>
                https://bugs.winehq.org/show_bug.cgi?id=50993
                https://bugs.winehq.org/attachment.cgi?id=70233
Co-authored-by: Eric Pouech <eric.pouech@gmail.com>
                https://bugs.winehq.org/show_bug.cgi?id=52386
                https://bugs.winehq.org/attachment.cgi?id=71626

bernhardu committed:
Improve HeapReAlloc/HEAP_REALLOC_IN_PLACE_ONLY.

ASAN_OPTIONS='verbosity=0:windows_hook_rtl_allocators=1' WINEDLLOVERRIDES="$F=n;*.dll=n" WINEDEBUG= wine64 z:/home/bernhard/data/entwicklung/2024/wine/wine-build/build-asan-pe/64/obj/dlls/urlmon/tests/x86_64-windows/urlmon_test.exe url

urlmon_test:url

ASAN_OPTIONS='verbosity=0:windows_hook_rtl_allocators=1' WINEDLLOVERRIDES="$F=n;*.dll=n" WINEDEBUG= wine64 z:/home/bernhard/data/entwicklung/2024/wine/wine-build/build-asan-pe/64/obj/dlls/urlmon/tests/x86_64-windows/urlmon_test.exe url
...
url.c:3528: Test marked todo: expected GetWindow_ICodeInstall
url.c:4077: file test (no callback)...
0130:err:secur32:establish_context alloc_buffer=00007FFFFE1FC510, alloc_buffer.pvBuffer=00007F331B4D0800 RtlGetLastWin32Error(): 0x0 :979
0130:err:secur32:establish_context alloc_buffer=00007FFFFE1FC510, alloc_buffer.pvBuffer=00007F331B4D0800 RtlGetLastWin32Error(): 0x0 :982
=================================================================
==300==ERROR: AddressSanitizer: attempting double-free on 0x7f331b4d0800 in thread T0:
022c:fixme:file:server_get_file_info Unsupported info class e
    #0 0x6ffffa60b113 in RtlFreeHeap /home/bernhard/data/entwicklung/2024/llvm-mingw/2024-10-18/llvm-mingw/llvm-project/compiler-rt\lib/asan/asan_malloc_win.cpp:444:3
    #1 0x6ffff346345f in FreeContextBuffer /home/bernhard/data/entwicklung/2024/wine/wine/dlls/secur32\secur32.c:651:5
    #2 0x6ffffc2b6cd3 in netcon_secure_connect_setup /home/bernhard/data/entwicklung/2024/wine/wine/dlls/wininet\netconnection.c:484:13
    #3 0x6ffffc2a8216 in NETCON_secure_connect /home/bernhard/data/entwicklung/2024/wine/wine/dlls/wininet\netconnection.c:612:11
    #4 0x6ffffc295941 in HTTP_HttpSendRequestW /home/bernhard/data/entwicklung/2024/wine/wine/dlls/wininet\http.c:5100:23
    #5 0x6ffffc29b999 in HttpSendRequestW /home/bernhard/data/entwicklung/2024/wine/wine/dlls/wininet\http.c:5759:8
    #6 0x6ffffc29b2ab in HttpSendRequestA /home/bernhard/data/entwicklung/2024/wine/wine/dlls/wininet\http.c:5799:14
    #7 0x0001400dfa52 in can_do_https /home/bernhard/data/entwicklung/2024/wine\wine/dlls/urlmon/tests/url.c:3987:11
    #8 0x0001400dfa52 in func_url /home/bernhard/data/entwicklung/2024/wine\wine/dlls/urlmon/tests/url.c:4080:12
    #9 0x00014011cd90 in run_test /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:765:5
    #10 0x00014011cd90 in main /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:884:12
    #11 0x00014011ec3f in mainCRTStartup /home/bernhard/data/entwicklung/2024/wine\wine/dlls/msvcrt/crt_main.c:58:11
    #12 0x6fffffa54808 in BaseThreadInitThunk /home/bernhard/data/entwicklung/2024/wine/wine/dlls/kernel32\thread.c:61:5
    #13 0x6fffffc2fa1a in RtlUserThreadStart (C:\windows\system32\ntdll.dll+0x17000fa1a)

0x7f331b4d0800 is located 0 bytes inside of 65536-byte region [0x7f331b4d0800,0x7f331b4e0800)
freed by thread T0 here:
    #0 0x6ffffa60abc6 in __asan::SharedReAlloc(void* (*)(void*, unsigned long, void*, unsigned long long), unsigned long long (*)(void*, unsigned long, void*), int (*)(void*, unsigned long, void*), void* (*)(void*, unsign
ed long, unsigned long long), void*, unsigned long, void*, unsigned long long) /home/bernhard/data/entwicklung/2024/llvm-mingw/2024-10-18/llvm-mingw/llvm-project/compiler-rt\lib/asan/asan_malloc_win.cpp:269:3
    #1 0x6ffffa60afa4 in HeapReAlloc /home/bernhard/data/entwicklung/2024/llvm-mingw/2024-10-18/llvm-mingw/llvm-project/compiler-rt\lib/asan/asan_malloc_win.cpp:394:10
    #2 0x6ffff346858d in establish_context.constprop.0 /home/bernhard/data/entwicklung/2024/wine/wine/dlls/secur32\schannel.c:980:13
    #3 0x6ffff3464680 in InitializeSecurityContextW /home/bernhard/data/entwicklung/2024/wine/wine/dlls/secur32\wrapper.c:249:19
    #4 0x6ffffc2b6bdf in netcon_secure_connect_setup /home/bernhard/data/entwicklung/2024/wine/wine/dlls/wininet\netconnection.c:464:14
    #5 0x6ffffc2a8216 in NETCON_secure_connect /home/bernhard/data/entwicklung/2024/wine/wine/dlls/wininet\netconnection.c:612:11
    #6 0x6ffffc295941 in HTTP_HttpSendRequestW /home/bernhard/data/entwicklung/2024/wine/wine/dlls/wininet\http.c:5100:23
    #7 0x6ffffc29b999 in HttpSendRequestW /home/bernhard/data/entwicklung/2024/wine/wine/dlls/wininet\http.c:5759:8
    #8 0x6ffffc29b2ab in HttpSendRequestA /home/bernhard/data/entwicklung/2024/wine/wine/dlls/wininet\http.c:5799:14
    #9 0x0001400dfa52 in can_do_https /home/bernhard/data/entwicklung/2024/wine\wine/dlls/urlmon/tests/url.c:3987:11
    #10 0x0001400dfa52 in func_url /home/bernhard/data/entwicklung/2024/wine\wine/dlls/urlmon/tests/url.c:4080:12
    #11 0x00014011cd90 in run_test /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:765:5
    #12 0x00014011cd90 in main /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:884:12
    #13 0x00014011ec3f in mainCRTStartup /home/bernhard/data/entwicklung/2024/wine\wine/dlls/msvcrt/crt_main.c:58:11
    #14 0x6fffffa54808 in BaseThreadInitThunk /home/bernhard/data/entwicklung/2024/wine/wine/dlls/kernel32\thread.c:61:5
    #15 0x6fffffc2fa1a in RtlUserThreadStart (C:\windows\system32\ntdll.dll+0x17000fa1a)

previously allocated by thread T0 here:
    #0 0x6ffffa60a8a3 in HeapAlloc /home/bernhard/data/entwicklung/2024/llvm-mingw/2024-10-18/llvm-mingw/llvm-project/compiler-rt\lib/asan/asan_malloc_win.cpp:230:3
    #1 0x6ffff3468098 in establish_context.constprop.0 /home/bernhard/data/entwicklung/2024/wine/wine/dlls/secur32\schannel.c:954:37
    #2 0x6ffff3464680 in InitializeSecurityContextW /home/bernhard/data/entwicklung/2024/wine/wine/dlls/secur32\wrapper.c:249:19
    #3 0x6ffffc2b6bdf in netcon_secure_connect_setup /home/bernhard/data/entwicklung/2024/wine/wine/dlls/wininet\netconnection.c:464:14
    #4 0x6ffffc2a8216 in NETCON_secure_connect /home/bernhard/data/entwicklung/2024/wine/wine/dlls/wininet\netconnection.c:612:11
    #5 0x6ffffc295941 in HTTP_HttpSendRequestW /home/bernhard/data/entwicklung/2024/wine/wine/dlls/wininet\http.c:5100:23
    #6 0x6ffffc29b999 in HttpSendRequestW /home/bernhard/data/entwicklung/2024/wine/wine/dlls/wininet\http.c:5759:8
    #7 0x6ffffc29b2ab in HttpSendRequestA /home/bernhard/data/entwicklung/2024/wine/wine/dlls/wininet\http.c:5799:14
    #8 0x0001400dfa52 in can_do_https /home/bernhard/data/entwicklung/2024/wine\wine/dlls/urlmon/tests/url.c:3987:11
    #9 0x0001400dfa52 in func_url /home/bernhard/data/entwicklung/2024/wine\wine/dlls/urlmon/tests/url.c:4080:12
    #10 0x00014011cd90 in run_test /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:765:5
    #11 0x00014011cd90 in main /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:884:12
    #12 0x00014011ec3f in mainCRTStartup /home/bernhard/data/entwicklung/2024/wine\wine/dlls/msvcrt/crt_main.c:58:11
    #13 0x6fffffa54808 in BaseThreadInitThunk /home/bernhard/data/entwicklung/2024/wine/wine/dlls/kernel32\thread.c:61:5
    #14 0x6fffffc2fa1a in RtlUserThreadStart (C:\windows\system32\ntdll.dll+0x17000fa1a)

SUMMARY: AddressSanitizer: double-free /home/bernhard/data/entwicklung/2024/wine/wine/dlls/secur32\secur32.c:651:5 in FreeContextBuffer
==300==ABORTING

bernhardu committed:
Fix checks for unsupports HEAP flags.

bernhardu committed:
[win/asan] GetInstructionSize: Support some more 5 byte instructions.

This patch adds several instructions seen when trying to run a
executable built with ASan with llvm-mingw.
(x86 and x86_64, using the git tip in llvm-project).

Also includes instructions collected by
Roman Pišl and Eric Pouech in the Wine bug reports below.

Related: https://github.com/llvm/llvm-project/issues/96270

Co-authored-by: Roman Pišl <rpisl@seznam.cz>
                https://bugs.winehq.org/show_bug.cgi?id=50993
                https://bugs.winehq.org/attachment.cgi?id=70233
Co-authored-by: Eric Pouech <eric.pouech@gmail.com>
                https://bugs.winehq.org/show_bug.cgi?id=52386
                https://bugs.winehq.org/attachment.cgi?id=71626

bernhardu committed:
[win/asan] GetInstructionSize: Support some more 6 byte instructions.

This patch adds several instructions seen when trying to run a
executable built with ASan with llvm-mingw.
(x86 and x86_64, using the git tip in llvm-project).

Also includes instructions collected by
Roman Pišl and Eric Pouech in the Wine bug reports below.

Related: https://github.com/llvm/llvm-project/issues/96270

Co-authored-by: Roman Pišl <rpisl@seznam.cz>
                https://bugs.winehq.org/show_bug.cgi?id=50993
                https://bugs.winehq.org/attachment.cgi?id=70233
Co-authored-by: Eric Pouech <eric.pouech@gmail.com>
                https://bugs.winehq.org/show_bug.cgi?id=52386
                https://bugs.winehq.org/attachment.cgi?id=71626

bernhardu committed:
6 byte comment

bernhardu committed:
4 byte test correction

bernhardu committed:
4 byte add `FF 74 24 XX`

bernhardu committed:
11/13 byte, reenable large NOPs

bernhardu committed:
TEMP: Avoid -gline-tables-only to get better debug information.

This is probably just important with `-DCMAKE_BUILD_TYPE=RelWithDebInfo`.

bernhardu committed:
TEMP: Leave optimization at -O2.

Just stumbled over following, and found a -O3 while looking for the
missing variables/parameter debug info, removed by -gline-tables-only.
Not sure if really needed.

llvm:CMakeLists.txt:1232
if( MINGW AND NOT "${CMAKE_CXX_COMPILER_ID}" MATCHES "Clang" )
  # People report that -O3 is unreliable on MinGW. The traditional
  # build also uses -O2 for that reason:
  llvm_replace_compiler_option(CMAKE_CXX_FLAGS_RELEASE "-O3" "-O2")
endif()

bernhardu committed:
Avoid memset in ASan error message printing.

`-ftrivial-auto-var-init=pattern` emits `memset` to initialize variables,
but with half-way intercepted functions this is problematic.

Can also be switched off per function with `__declspec(no_init_all)`.



Program received signal SIGSEGV, Segmentation fault.
0x7a80f51f in __asan_wrap_memset (dst=0x121ed78, v=255, size=400) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors_memintrinsics.inc:87
87        COMMON_INTERCEPTOR_MEMSET_IMPL(ctx, dst, v, size);
1: x/i $pc
=> 0x7a80f51f <__asan_wrap_memset(void*, int, __sanitizer::usize)+335>: push   %eax
(rr) bt 10
#0  0x7a80f51f in __asan_wrap_memset (dst=0x121ed78, v=255, size=400) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors_memintrinsics.inc:87
#1  0x7a80f44b in __asan_wrap_memset (dst=0x121ed78, v=255, size=400) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors_memintrinsics.inc:87
#2  0x7a80f44b in __asan_wrap_memset (dst=0x121ed78, v=255, size=400) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors_memintrinsics.inc:87
#3  0x7a80f44b in __asan_wrap_memset (dst=0x121ed78, v=255, size=400) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors_memintrinsics.inc:87
...
#9168 0x7a80f44b in __asan_wrap_memset (dst=0x121ed78, v=255, size=400) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors_memintrinsics.inc:87
#9169 0x7a80f44b in __asan_wrap_memset (dst=0x121ed78, v=255, size=400) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors_memintrinsics.inc:87
#9170 0x7a80f44b in __asan_wrap_memset (dst=0x121ed78, v=255, size=400) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors_memintrinsics.inc:87
#9171 0x7a7db0c7 in __sanitizer::SharedPrintfCode (append_pid=false, format=0x7a82b160 <.refptr.___ImageBase+4> "%s: CHECK failed: %s:%d \"%s\" (0x%zx, 0x%zx) (tid=%u)\n", args=<optimized out>) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/sanitizer_common/sanitizer_printf.cpp:308
#9172 0x0000000c in ?? ()


(rr) disassemble __sanitizer::SharedPrintfCode
Dump of assembler code for function _ZN11__sanitizerL16SharedPrintfCodeEbPKcPc:
   0x7a7db0a0 <+0>:     push   %ebp
   0x7a7db0a1 <+1>:     push   %ebx
   0x7a7db0a2 <+2>:     push   %edi
   0x7a7db0a3 <+3>:     push   %esi
   0x7a7db0a4 <+4>:     sub    $0x190,%esp
   0x7a7db0aa <+10>:    mov    %edx,%esi
   0x7a7db0ac <+12>:    mov    %ecx,%edi
   0x7a7db0ae <+14>:    mov    0x1a4(%esp),%ebp
   0x7a7db0b5 <+21>:    mov    %esp,%ebx
   0x7a7db0b7 <+23>:    push   $0x190
   0x7a7db0bc <+28>:    push   $0xff
   0x7a7db0c1 <+33>:    push   %ebx
=> 0x7a7db0c2 <+34>:    call   0x7a823bc6 <memset>
   0x7a7db0c7 <+39>:    add    $0xc,%esp
   0x7a7db0ca <+42>:    mov    %edi,%ecx
   0x7a7db0cc <+44>:    mov    %ebx,%edx
   0x7a7db0ce <+46>:    push   %ebp
   0x7a7db0cf <+47>:    push   %esi
   0x7a7db0d0 <+48>:    call   0x7a7db5c0 <_ZN11__sanitizerL24SharedPrintfCodeNoBufferEbPciPKcS0_>
   0x7a7db0d5 <+53>:    add    $0x198,%esp
   0x7a7db0db <+59>:    pop    %esi
   0x7a7db0dc <+60>:    pop    %edi
   0x7a7db0dd <+61>:    pop    %ebx
   0x7a7db0de <+62>:    pop    %ebp
   0x7a7db0df <+63>:    ret
End of assembler dump.


With the `no_init_all` ASan prints:
AddressSanitizer: CHECK failed: asan_suppressions.cpp:46 "((suppression_ctx)) != (0)" (0x0, 0x0) (tid=36)


Unfortunately it looks like this draws in almost all dlls ...

                    /* llvm-symbolizer.exe, avoid loading libclang_rt.asan_dynamic-i386.dll too early */
                    strstr(make->module, "msvcp140.dll") ||
                    strstr(make->module, "vcruntime140.dll") ||
                    strstr(make->module, "concrt140.dll") ||
                    strstr(make->module, "advapi32.dll") ||
                    strstr(make->module, "sechost.dll") ||
                    strstr(make->module, "gdi32.dll") ||
                    strstr(make->module, "oleaut32.dll") ||
                    strstr(make->module, "imm32.dll") ||
                    strstr(make->module, "rpcrt4.dll") ||
                    strstr(make->module, "combase.dll") ||
                    strstr(make->module, "ole32.dll") ||
                    strstr(make->module, "iphlpapi.dll") ||
                    strstr(make->module, "dnsapi.dll") ||
                    strstr(make->module, "netapi32.dll") ||
                    strstr(make->module, "nsi.dll") ||
                    strstr(make->module, "ws2_32.dll") ||

                    strstr(make->module, "winspool.drv") ||
                    strstr(make->module, "compstui.dll") ||
                    strstr(make->module, "comctl32.dll") ||
                    strstr(make->module, "comctl32.dll") ||

and more ... but it looked working with x86_64 ?

bernhardu committed:
TEMP: Try to get better backtraces with ASan.

bernhardu committed:
Workaround: Avoid failing to calloc while half-way through AsanInitInternal.

- Get atexit table initialized before we intall interception.
- Do InitializeSuppressions a little earlier


(With LLVM-19.1.6-win32.exe)
ASAN_OPTIONS=verbosity=2 wine "C:\\Program Files\\LLVM\\bin\\llvm-symbolizer.exe"

AddressSanitizer: CHECK failed: asan_suppressions.cpp:46 "((suppression_ctx)) != (0)" (0x0, 0x0) (tid=324)
AddressSanitizer: CHECK failed: asan_suppressions.cpp:46 "((suppression_ctx)) != (0)" (0x0, 0x0) (tid=324)

(rr) bt
#0  __sanitizer::CheckFailed (file=0x7a892243 <.refptr.___asan_shadow_memory_dynamic_address+491> "/home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/asan_suppressions.cpp", line=46, cond=0x7a8922e4 <.refptr.___asan_shadow_memory_dynamic_address+652> "((suppression_ctx)) != (0)", v1=0, v2=0) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/sanitizer_common/sanitizer_termination.cpp:83
#1  0x7a8772bc in __asan::IsInterceptorSuppressed (interceptor_name=0x7a890364 <.refptr.___asan_option_detect_stack_use_after_return+9824> "memset") at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/asan_suppressions.cpp:46
#2  0x7a86c071 in __asan_wrap_memset (dst=0x302ba000, v=254, size=8192) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors_memintrinsics.inc:87
#3  0x7a86e563 in __asan::FastPoisonShadow (aligned_beg=<optimized out>, aligned_size=<optimized out>, value=<optimized out>) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/asan_poisoning.h:55
#4  __asan::PoisonShadow (addr=22872064, size=65536, value=254 '\376') at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/asan_poisoning.cpp:43
#5  0x7a875c43 in __asan::OnLowLevelAllocate (ptr=22872064, size=65536) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/asan_rtl.cpp:106
#6  0x7a832835 in __sanitizer::LowLevelAllocator::Allocate (this=0x7b14d138, size=16400) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/sanitizer_common/sanitizer_allocator.cpp:159
#7  0x7a84d3ba in operator new (size=16396, alloc=...) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/sanitizer_common/sanitizer_common.h:1112
#8  __sanitizer::ChooseSymbolizerTools (list=<optimized out>, allocator=<optimized out>) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/sanitizer_common/sanitizer_symbolizer_win.cpp:299
#9  __sanitizer::Symbolizer::PlatformInit () at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/sanitizer_common/sanitizer_symbolizer_win.cpp:312
#10 0x7a849834 in __sanitizer::Symbolizer::GetOrInit () at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/sanitizer_common/sanitizer_symbolizer_libcdep.cpp:24
#11 0x7a848378 in __sanitizer::(anonymous namespace)::StackTraceTextPrinter::ProcessAddressFrames (this=this@entry=0x121ead4, pc=2055691253) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/sanitizer_common/sanitizer_stacktrace_libcdep.cpp:37
#12 0x7a8482c1 in __sanitizer::StackTrace::PrintTo (this=this@entry=0x121eb30, output=0x121eb14) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/sanitizer_common/sanitizer_stacktrace_libcdep.cpp:110
#13 0x7a8485e1 in __sanitizer::StackTrace::Print (this=0x121eb30) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/sanitizer_common/sanitizer_stacktrace_libcdep.cpp:133
#14 0x7a875c21 in __asan::CheckUnwind () at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/asan_rtl.cpp:70
#15 0x7a843f03 in __sanitizer::CheckFailed (file=0x7a892243 <.refptr.___asan_shadow_memory_dynamic_address+491> "/home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/asan_suppressions.cpp", line=46, cond=0x7a8922e4 <.refptr.___asan_shadow_memory_dynamic_address+652> "((suppression_ctx)) != (0)", v1=0, v2=0) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/sanitizer_common/sanitizer_termination.cpp:86
#16 0x7a8772bc in __asan::IsInterceptorSuppressed (interceptor_name=0x7a890364 <.refptr.___asan_option_detect_stack_use_after_return+9824> "memset") at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/asan_suppressions.cpp:46
#17 0x7a86c071 in __asan_wrap_memset (dst=0x30300000, v=250, size=131072) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors_memintrinsics.inc:87
#18 0x7a86e563 in __asan::FastPoisonShadow (aligned_beg=<optimized out>, aligned_size=<optimized out>, value=<optimized out>) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/asan_poisoning.h:55
#19 __asan::PoisonShadow (addr=25165824, size=1048576, value=250 '\372') at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/asan_poisoning.cpp:43
#20 0x7a856c54 in __asan::AsanMapUnmapCallback::OnMap (p=25165824, size=1048576, this=<optimized out>) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/asan_allocator.cpp:283
#21 __sanitizer::SizeClassAllocator32<__asan::AP32<__sanitizer::LocalAddressSpaceView> >::AllocateRegion (stat=0x7b1777e4, this=<optimized out>, class_id=<optimized out>) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/../sanitizer_common/sanitizer_allocator_primary32.h:304
#22 __sanitizer::SizeClassAllocator32<__asan::AP32<__sanitizer::LocalAddressSpaceView> >::PopulateFreeList (this=0x7b14ec40, stat=0x7b1777e4, c=0x7b170be4, sci=0x7b14ff40, class_id=12) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/../sanitizer_common/sanitizer_allocator_primary32.h:343
#23 0x7a856b97 in __sanitizer::SizeClassAllocator32<__asan::AP32<__sanitizer::LocalAddressSpaceView> >::AllocateBatch (this=<optimized out>, stat=0x7b1777e4, c=0x7b170be4, class_id=12) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/../sanitizer_common/sanitizer_allocator_primary32.h:174
#24 0x7a856a2b in __sanitizer::SizeClassAllocator32LocalCache<__sanitizer::SizeClassAllocator32<__asan::AP32<__sanitizer::LocalAddressSpaceView> > >::Refill (this=this@entry=0x7b170be4, c=0x7b1723e4, allocator=0x7b14ec40, class_id=12) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/../sanitizer_common/sanitizer_allocator_local_cache.h:244
#25 0x7a8575f5 in __sanitizer::SizeClassAllocator32LocalCache<__sanitizer::SizeClassAllocator32<__asan::AP32<__sanitizer::LocalAddressSpaceView> > >::Allocate (this=0x7b170be4, allocator=0x7b14ec40, class_id=131072) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/../sanitizer_common/sanitizer_allocator_local_cache.h:164
#26 __sanitizer::CombinedAllocator<__sanitizer::SizeClassAllocator32<__asan::AP32<__sanitizer::LocalAddressSpaceView> >, __sanitizer::LargeMmapAllocatorPtrArrayStatic>::Allocate (this=this@entry=0x7b14ec40, cache=0x7b170be4, size=<optimized out>, alignment=8) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/../sanitizer_common/sanitizer_allocator_combined.h:69
#27 0x7a85545b in __asan::Allocator::Allocate (this=<optimized out>, size=<optimized out>, alignment=8, stack=0x121f5c0, alloc_type=__asan::FROM_MALLOC, can_fill=<optimized out>) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/asan_allocator.cpp:586
#28 0x7a853aee in __asan::Allocator::Calloc (nmemb=32, size=4, stack=0x121f5c0, this=<optimized out>) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/asan_allocator.cpp:787
#29 __asan::asan_calloc (nmemb=32, size=4, stack=0x121f5c0) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/asan_allocator.cpp:1012
#30 0x7a86cd43 in calloc (nmemb=32, size=4) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/asan_malloc_win.cpp:92
#31 0x7b493543 in register_onexit_function (table=table@entry=0x7a89b160 <atexit_table>, func=func@entry=0x7a844430 <__sanitizer_cov_dump()>) at /home/bernhard/data/entwicklung/2024/wine/wine/dlls/msvcrt/exit.c:80
#32 0x7b493a54 in _register_onexit_function (table=0x7a89b160 <atexit_table>, func=0x7a844430 <__sanitizer_cov_dump()>) at /home/bernhard/data/entwicklung/2024/wine/wine/dlls/msvcrt/exit.c:443
#33 0x7a83139e in atexit (func=<optimized out>) at ../crt/crtdll.c:209
#34 0x7a841f6d in __sanitizer::Atexit (function=0x7a844430 <__sanitizer_cov_dump()>) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/sanitizer_common/sanitizer_win.cpp:752
#35 0x7a84440f in __sanitizer::InitializeCoverage (enabled=64, dir=0x7a87f9f6 <.refptr.__ZN11__sanitizer24stoptheworld_tracer_ppidE+42> ".") at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/sanitizer_common/sanitizer_coverage_libcdep_new.cpp:214
#36 0x7a844430 in __sanitizer::InitializeCoverage(bool, char const*) () at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/sanitizer_common/sanitizer_coverage_libcdep_new.cpp:216
#37 0x0121fa7c in ?? ()
#38 0x7a8757ab in __asan::AsanInitInternal () at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/asan_rtl.cpp:474
#39 0x7a875fa0 in __asan::AsanInitFromRtl () at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/asan_rtl.cpp:522
#40 __asan::AsanInitializer::AsanInitializer (this=<optimized out>) at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/asan_rtl.cpp:542
#41 __cxx_global_var_init () at /home/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/lib/asan/asan_rtl.cpp:546
#42 0x7a875fa0 in _GLOBAL__sub_I_asan_rtl.cpp () from /media/bernhard/data/entwicklung/2025/llvm-mingw/2025-01-04/llvm-mingw/llvm-project/compiler-rt/build-i686-sanitizers/lib/windows/libclang_rt.asan_dynamic-i386.dll
#43 0x0121fd24 in ?? ()
#44 0x7a87cb9b in __do_global_ctors () at ../crt/gccmain.c:44
#45 __main () at ../crt/gccmain.c:58
#46 0x7a831222 in __DllMainCRTStartup (hDllHandle=0x7a830000, dwReason=1, lpreserved=<optimized out>) at ../crt/crtdll.c:185
#47 0x7a8311af in _DllMainCRTStartup@12 (hDllHandle=<optimized out>, dwReason=<optimized out>, lpreserved=<optimized out>) at ../crt/crtdll.c:156
#48 0x7bdf63d6 in call_dll_entry_point () from /mnt/sdb1/wineprefix_asan_2025-01-01_win32/drive_c/windows/system32/ntdll.dll
#49 0x7bdf96e0 in MODULE_InitDLL (wm=<optimized out>, wm@entry=0x243bd0, reason=reason@entry=1, lpReserved=0x121fd24) at /home/bernhard/data/entwicklung/2024/wine/wine/dlls/ntdll/loader.c:1668
#50 0x7bdf7b42 in process_attach (node=<optimized out>, lpReserved=lpReserved@entry=0x121fd24) at /home/bernhard/data/entwicklung/2024/wine/wine/dlls/ntdll/loader.c:1762
#51 0x7bdfa4d8 in loader_init (context=0x121fd24, entry=0x121fdd4) at /home/bernhard/data/entwicklung/2024/wine/wine/dlls/ntdll/loader.c:4456
#52 0x7be18b2c in _LdrInitializeThunk@16 (context=0x121fd24, unk2=0, unk3=0, unk4=0) at /home/bernhard/data/entwicklung/2024/wine/wine/dlls/ntdll/signal_i386.c:559
#53 0xdeadbabe in ?? ()

bernhardu committed:
llvm-mingw: Enable tests and improve debugging.

bernhardu commented:
I guess this PR serves no purpose any longer, so I am going to close it.

Thanks for you time, the important patches for GetInstructionSize went in.
Gladly before the branch of LLVM 20, was not aware this was so near, so thanks again for getting them merged in time.
